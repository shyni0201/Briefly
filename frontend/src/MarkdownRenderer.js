/**
 * MarkdownRenderer Component
 * Renders markdown content with support for syntax highlighting and GitHub-Flavored Markdown (GFM).
 * 
 * @component
 * @param {object} props - Component props
 * @param {string} props.content - Markdown content to render
 */
import React from 'react'; // Core React library
import ReactMarkdown from 'react-markdown'; // Library to render markdown as React components
import remarkGfm from 'remark-gfm'; // Plugin for GitHub-Flavored Markdown (GFM), supporting tables, strikethroughs, etc.
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter'; // Syntax highlighter for code blocks
import { oneDark } from 'react-syntax-highlighter/dist/esm/styles/prism'; // Theme for syntax highlighting

/**
 * MarkdownRenderer Component
 * - Utilizes `ReactMarkdown` for rendering markdown.
 * - Supports GitHub-Flavored Markdown (GFM) with the `remark-gfm` plugin.
 * - Implements syntax highlighting for code blocks using `react-syntax-highlighter`.
 *
 * @param {string} content - Markdown content to be rendered.
 */
const MarkdownRenderer = ({ content }) => {
  return (
    /**
     * ReactMarkdown component renders markdown content as React elements.
     * - `remarkPlugins`: Configures plugins like `remark-gfm` for additional markdown features.
     * - `components`: Customizes the rendering of specific markdown elements (e.g., `code` blocks, paragraphs).
     */
    <ReactMarkdown
      remarkPlugins={[remarkGfm]} // Enable GitHub-flavored Markdown
      components={{
         /**
         * Custom rendering for `code` blocks:
         * - Uses `react-syntax-highlighter` for syntax highlighting of code blocks.
         * - Matches the language from the `className` (e.g., `language-js` for JavaScript).
         * - If inline code, renders it within a `<code>` tag without highlighting.
         *
         * @param {object} params - Props passed to the code block.
         * @param {boolean} params.inline - Indicates whether the code block is inline.
         * @param {string} params.className - Class name of the code block (used to determine language).
         * @param {Array} params.children - Child nodes (actual code content).
         */
        code({ node, inline, className, children, ...props }) {
          const match = /language-(\w+)/.exec(className || '');
          return !inline && match ? (
            <SyntaxHighlighter
              style={oneDark} // Syntax highlighting theme
              language={match[1]} // Language for syntax highlighting (e.g., `javascript`, `python`)
              PreTag="div" // Wrapper tag for the highlighted code block
              {...props} // Spread additional props
            >
              {String(children).replace(/\n$/, '')} {/* Render code content */}
            </SyntaxHighlighter>
          ) : (
            <code
              className={className}
              {...props}
              style={{
                whiteSpace: 'pre-wrap',
                wordBreak: 'break-word',
                overflowWrap: 'break-word',
                background: '#f5f5f5',
                padding: '2px 4px',
                borderRadius: '4px',
                lineHeight: '1.6',
                fontSize: '16px',
              }}
            >
              {children}
            </code>
          );
        },
        /**
         * Custom rendering for `p` (paragraph) elements:
         * - Adds consistent margin and font size for paragraphs.
         *
         * @param {object} params - Props passed to the paragraph element.
         * @param {Array} params.children - Child nodes (text content).
         */
        p({ children }) {
          return (
            <p
              style={{
                margin: '8px 0', // Adjust top and bottom margin
                fontSize: '16px', // Font size for readability
                lineHeight: '1.6', // Adjust line height for better spacing
              }}
            >
              {children}
            </p>
          );
        },
      }}
    >
      {content}
    </ReactMarkdown>
  );
};

export default MarkdownRenderer; // Export the component for use in other parts of the application