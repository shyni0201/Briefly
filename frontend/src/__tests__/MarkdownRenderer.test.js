import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import MarkdownRenderer from '../MarkdownRenderer';

jest.mock('react-syntax-highlighter', () => ({
  Prism: ({ children }) => <div data-testid="syntax-highlighter">{children}</div>,
}));

jest.mock('react-syntax-highlighter/dist/esm/styles/prism', () => ({
  oneDark: {},
}));

jest.mock('react-markdown', () => (props) => (
    <div data-testid="react-markdown">{props.children}</div>
  ));

jest.mock('remark-gfm', () => jest.fn());

describe('MarkdownRenderer Component', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        jest.resetModules();
      });
  test('renders plain text content', () => {
    const plainText = 'This is a plain text paragraph.';
    render(<MarkdownRenderer content={plainText} />);

    expect(screen.getByText(plainText)).toBeInTheDocument();
  });

  test('renders headings correctly', () => {
    const headings = '# Heading 1\n## Heading 2\n### Heading 3';
    render(<MarkdownRenderer content={headings} />);

    expect(screen.getByTestId('react-markdown')).toBeInTheDocument();
    expect(screen.getByTestId('react-markdown')).toHaveTextContent('# Heading 1');
    expect(screen.getByTestId('react-markdown')).toHaveTextContent('## Heading 2');
    expect(screen.getByTestId('react-markdown')).toHaveTextContent('### Heading 3');
  });

  test('renders Markdown formatted content', () => {
    const markdownContent = '# Heading 1\n\nThis is a paragraph.';
    render(<MarkdownRenderer content={markdownContent} />);
    const codeBlock = screen.getByTestId('react-markdown');
    expect(screen.getByTestId('react-markdown')).toBeInTheDocument();
    expect(codeBlock).toBeInTheDocument();
  });

  test('renders a code block with syntax highlighting', () => {
    const codeContent = '```js\nconsole.log("Hello, world!");\n```';
    render(<MarkdownRenderer content={codeContent} />);

    expect(screen.getByTestId('react-markdown')).toBeInTheDocument();
    expect(screen.getByTestId('react-markdown')).toHaveTextContent('console.log("Hello, world!");');
  });

  test('renders nested lists', () => {
    const nestedList = `
- Item 1
  - Subitem 1.1
  - Subitem 1.2
- Item 2
`;
    render(<MarkdownRenderer content={nestedList} />);
    expect(screen.getByTestId('react-markdown')).toHaveTextContent('Item 1');
    expect(screen.getByTestId('react-markdown')).toHaveTextContent('Subitem 1.1');
    expect(screen.getByTestId('react-markdown')).toHaveTextContent('Item 2');
  });

  test('renders inline code', () => {
    const inlineCode = 'This is `inline code`.';
    render(<MarkdownRenderer content={inlineCode} />);
    const codeBlock = screen.getByTestId('react-markdown');
    expect(screen.getByTestId('react-markdown')).toBeInTheDocument();
    expect(codeBlock).toBeInTheDocument();
  });

  test('renders fallback for inline code without language', () => {
    const inlineCode = '`No language specified`';
    render(<MarkdownRenderer content={inlineCode} />);
    const codeBlock = screen.getByTestId('react-markdown');
    expect(codeBlock).toBeInTheDocument();
    expect(codeBlock).toHaveTextContent('No language specified');
  });

  test('renders non-inline code block with syntax highlighting', () => {
    const codeContent = '```js\nconsole.log("Hello, world!");\n```';
    render(<MarkdownRenderer content={codeContent} />);
  
    const syntaxHighlighter = screen.getByTestId('react-markdown');
    expect(syntaxHighlighter).toBeInTheDocument();
    expect(syntaxHighlighter).toHaveTextContent('console.log("Hello, world!");');
  });

  test('renders GitHub-flavored Markdown (GFM)', () => {
    const gfmContent = '- [x] Task 1\n- [ ] Task 2\n\n| Col1 | Col2 |\n| ---- | ---- |\n| Val1 | Val2 |';
    render(<MarkdownRenderer content={gfmContent} />);
    const codeBlock = screen.getByTestId('react-markdown');
    expect(screen.getByTestId('react-markdown')).toBeInTheDocument();
    expect(codeBlock).toBeInTheDocument();
  });

  test('renders a code block without language correctly', () => {
    const codeContent = '```\nconsole.log("No language specified");\n```';
    render(<MarkdownRenderer content={codeContent} />);
    const codeBlock = screen.getByTestId('react-markdown');
    expect(codeBlock).toBeInTheDocument();
    expect(codeBlock).toHaveTextContent('console.log("No language specified");');
  });

  test('renders custom styles for paragraphs', () => {
    const paragraph = 'This is a styled paragraph.';
    render(<MarkdownRenderer content={paragraph} />);
    const codeBlock = screen.getByTestId('react-markdown');
    expect(screen.getByTestId('react-markdown')).toBeInTheDocument();
    expect(codeBlock).toBeInTheDocument();
  });

  test('renders without crashing for empty content', () => {
    render(<MarkdownRenderer content="" />);
    expect(screen.getByTestId('react-markdown')).toBeInTheDocument();
  });

  test('renders Markdown content with tables', () => {
    const tableContent = `
| Header 1 | Header 2 |
|----------|----------|
| Cell 1   | Cell 2   |
`;
    render(<MarkdownRenderer content={tableContent} />);
    const renderedMarkdown = screen.getByTestId('react-markdown');
    expect(renderedMarkdown).toBeInTheDocument();
    expect(renderedMarkdown).toHaveTextContent('Header 1');
    expect(renderedMarkdown).toHaveTextContent('Cell 2');
  });

  test('matches the snapshot for complex Markdown content', () => {
    const complexContent = `
# Heading 1

This is a paragraph with **bold** and *italic* text.

\`\`\`js
console.log("Hello, world!");
\`\`\`

- Item 1
- Item 2
- Item 3

| Header 1 | Header 2 |
|----------|----------|
| Cell 1   | Cell 2   |
`;
    const { asFragment } = render(<MarkdownRenderer content={complexContent} />);
    expect(asFragment()).toMatchSnapshot();
  });
});