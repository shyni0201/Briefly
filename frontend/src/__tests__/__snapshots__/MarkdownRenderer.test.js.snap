// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownRenderer Component matches the snapshot for complex Markdown content 1`] = `
<DocumentFragment>
  <div
    data-testid="react-markdown"
  >
    
# Heading 1

This is a paragraph with **bold** and *italic* text.

\`\`\`js
console.log("Hello, world!");
\`\`\`

- Item 1
- Item 2
- Item 3

| Header 1 | Header 2 |
|----------|----------|
| Cell 1   | Cell 2   |

  </div>
</DocumentFragment>
`;
